compiler/rustc_trait_selection/src/solve/normalize.rs

```diff
@@ -253,20 +253,28 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for DeeplyNormalizeForDiagnosticsFolder<'_,
     }
 
     fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
-        deeply_normalize_with_skipped_universes(
-            self.at,
-            ty,
-            vec![None; ty.outer_exclusive_binder().as_usize()],
-        )
-        .unwrap_or_else(|_: Vec<ScrubbedTraitError<'tcx>>| ty.super_fold_with(self))
+        self.at
+            .infcx
+            .commit_if_ok(|_| {
```

Comment:
`commit_if_ok` b/c we shouldn't apply inference/opaque side-effects if the nested goals don't hold.